Back-port of the following upstream commit...

commit a57a59f247b651e8ed6d3eeb7e2f9d83b83134c9
Author: Johannes Berg <johannes@sipsolutions.net>
Date:   Tue Oct 28 18:21:05 2008 +0100

    iwlwifi: remove implicit direct scan

    When an undirected scan is requested and iwlwifi is not associated but
    the user has set an SSID (and maybe was associated with that network at
    some point) then iwlwifi will assume the user wanted to scan for this
    SSID which seems wrong. Remove this code.

    Signed-off-by: Johannes Berg <johannes@sipsolutions.net>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff -up linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl3945-base.c.orig linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl3945-base.c
--- linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl3945-base.c.orig	2009-04-07 14:34:39.000000000 -0400
+++ linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl3945-base.c	2009-04-07 14:43:08.000000000 -0400
@@ -6202,14 +6202,6 @@ static void iwl3945_bg_request_scan(stru
 		memcpy(scan->direct_scan[0].ssid,
 		       priv->direct_ssid, priv->direct_ssid_len);
 		direct_mask = 1;
-	} else if (!iwl3945_is_associated(priv) && priv->essid_len) {
-		IWL_DEBUG_SCAN
-		  ("Kicking off one direct scan for '%s' when not associated\n",
-		   iwl3945_escape_essid(priv->essid, priv->essid_len));
-		scan->direct_scan[0].id = WLAN_EID_SSID;
-		scan->direct_scan[0].len = priv->essid_len;
-		memcpy(scan->direct_scan[0].ssid, priv->essid, priv->essid_len);
-		direct_mask = 1;
 	} else {
 		IWL_DEBUG_SCAN("Kicking off one indirect scan.\n");
 		direct_mask = 0;
diff -up linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl-scan.c.orig linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl-scan.c
--- linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl-scan.c.orig	2009-04-07 14:34:27.000000000 -0400
+++ linux-2.6.27.noarch/drivers/net/wireless/iwlwifi/iwl-scan.c	2009-04-07 14:42:12.000000000 -0400
@@ -804,13 +804,6 @@ static void iwl_bg_request_scan(struct w
 		memcpy(scan->direct_scan[0].ssid,
 		       priv->direct_ssid, priv->direct_ssid_len);
 		n_probes++;
-	} else if (!iwl_is_associated(priv) && priv->essid_len) {
-		IWL_DEBUG_SCAN("Start direct scan for '%s' (not associated)\n",
-				iwl_escape_essid(priv->essid, priv->essid_len));
-		scan->direct_scan[0].id = WLAN_EID_SSID;
-		scan->direct_scan[0].len = priv->essid_len;
-		memcpy(scan->direct_scan[0].ssid, priv->essid, priv->essid_len);
-		n_probes++;
 	} else {
 		IWL_DEBUG_SCAN("Start indirect scan.\n");
 	}
